aesara/tensor/math_opt.py

```diff
@@ -1862,6 +1862,71 @@ def local_neg_div_neg(fgraph, node):
                     return [true_div(new_num, denom)]
 
 
+@register_canonicalize
+@local_optimizer([sub])
+def local_sub_neg_to_add(fgraph, node):
+    """
+    x - (-y) -> x + y
+    """
+
+    if node.op == sub:
+        otype = node.outputs[0].dtype
+        minuend, subtrahend = node.inputs
+
+        if subtrahend.owner:
+            if subtrahend.owner.op == neg:
+                pre_neg = subtrahend.owner.inputs[0]
+                new_out = add(minuend, pre_neg)
+                if new_out.dtype != otype:
+                    new_out = cast(new_out, dtype=otype)
+                return [new_out]
+
+            # Check if negation is hidden behind a DimShuffle
+            elif isinstance(subtrahend.owner.op, DimShuffle):
```

Comment:
I'm always very wary of these compound canonicalization steps.  Is there an independent canonicalization we're missing that would get rid of this `DimShuffle`?  If so, let's add that instead.

Reply:
You mean like pushing the dimshuffle inside the`neg`? That would be a less efficient graph when we are not removing the `neg`, no?

Or would we revert it in a later stage?

Looking at other rewrites in that file, more often than not there is a TODO comment saying the rewrite should also have looked for dimshuffles. So finding a general solution would probably be useful beyond these new rewrites.

Reply:
> That would be a less efficient graph when we are not removing the `neg`, no?

In what sense?

Reply:
That applying the `neg` after a dimshuffle would be slower than applying it before.

Reply:
Canonicalization isn't about producing a graph that evaluates faster; those kinds of rewrites are applied afterward.